core/src/main/scala/org/genivi/sota/core/CampaignResource.scala

```diff
@@ -47,9 +49,17 @@ class CampaignResource(namespaceExtractor: Directive1[AuthedNamespaceScope],
     complete(db.run(Campaigns.fetch(id)))
   }
 
-  def launch(id: Campaign.Id, lc: LaunchCampaign): Route = {
-    lc.isValid match {
-      case Xor.Right(()) => complete(CampaignLauncher.launch(deviceRegistry, updateService, id, lc, messageBus))
+  def launch(id: Campaign.Id, lc: LaunchCampaignRequest): Route = {
+    lc.isValid() match {
+      case Xor.Right(()) =>
+        val r = async {
+          val campaign = await(db.run(Campaigns.fetch(id)))
+          campaign.meta.deltaFrom match {
+            case Some(_) => await(CampaignLauncher.generateDeltaRequest(id, lc, messageBus))
+            case None => await(CampaignLauncher.launch(deviceRegistry, updateService, id, lc, messageBus))
```

Comment:
Both of `CampaignLauncher.generateDeltaRequest` and `CampaignLauncher.launch` will start with `db.run(Campaigns.fetch(id))`, so we will lookup this twice, could we instead make case distinction in `launch` and have the `generateDeltaRequest` be private to `CampaignLauncher`? It could take a `Campaign` object instead of an id as well.